_posts/2023-09-11-Rails-7-1-0-beta-1-has-been-released.markdown

```diff
@@ -0,0 +1,143 @@
+is intended to be used in the controller to authenticate the user, and it is implemented in a way that protect against
+common timing attacks.
+
+```ruby
+User.authenticate_by(email: "rafael@rubyonrails.org", password: "railsrocks")
+```
+
+[`generates_token_for`](https://edgeapi.rubyonrails.org/classes/ActiveRecord/TokenFor/ClassMethods.html#method-i-generates_token_for)
+can be used to implement features like password reset, email confirmation, and other features that require a single use
+token.
+
+```ruby
+class User < ApplicationRecord
+  has_secure_password
+
+  generates_token_for :password_reset, expires_in: 15.minutes do
+    # Last 10 characters of password salt, which changes when password is updated:
+    password_salt&.last(10)
+  end
+end
+
+user = User.first
+
+token = user.generate_token_for(:password_reset)
+User.find_by_token_for(:password_reset, token) # => user
+# 16 minutes later...
+User.find_by_token_for(:password_reset, token) # => nil
+```
+
+Finally, [`has_secure_password`](https://edgeapi.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password)
+got the ability to automatically verify the current password when updating the password. This is useful to check
+if the user who is trying to update the password, knows the current password.
+
+```ruby
+# Schema: User(name:string, password_digest:string)
+class User < ActiveRecord::Base
+  has_secure_password
+end
+
+user = User.new(name: "rafael", password: "railsrocks", password_confirmation: "railsrocks")
+user.save                                                                      # => true
+user.update(password: "pwn3d", password_challenge: "")                         # => false, challenge doesn't authenticate
+user.update(password: "railsGetsEvenBetter", password_challenge: "railsrocks") # => true
+```
+
+## A more robust framework
+
+Rails promises to be a framework that can get you from Hello World to IPO, and on each release we are improving
+how robust Rails is.
+
+For 7.1 we are bringing a collection of changes that have the goal of making easier for your application to scale.
+
+The [`perform_all_later` method in Active Job](https://edgeapi.rubyonrails.org/classes/ActiveJob.html#method-c-perform_all_later),
+was designed to streamline the process of enqueuing multiple jobs simultaneously. This powerful
+addition allows you to efficiently enqueue jobs without triggering callbacks. This is
+particularly useful when you need to enqueue a batch of jobs at once, reducing the overhead
+of multiple round-trips to the queue datastore.
+
+Support to Composite Primary keys was added to Active Record, [allowing applications to improve the performance of some
+queries](https://shopify.engineering/how-to-introduce-composite-primary-keys-in-rails), specially those applications
+that use multi-tenant architecture, by using the tenant id as part of the primary key.
```

Comment:
```suggestion
that use multi-tenant architecture, by using the tenant ID as part of the primary key.
```

or

```suggestion
that use multi-tenant architecture, by using the tenant `id` as part of the primary key.
```

Reply:
Agree with this suggestion! I'd say the second option could be `tenant_id` if we want to imply a particular column name or "tenant ID" if we want to imply "tenant identifier"